src/common/smartPay.ts

```diff
@@ -0,0 +1,475 @@
+import { EdgeCurrencyInfo, EdgeIo, EdgeParsedUri } from 'edge-core-js'
+
+import { cleanMultiFetch, makeQueryParams, QueryParams } from './network'
+
+const MAX_TIMEOUT_S = 60 * 60 * 24 * 7
+
+const asSmartPayQrDecode = asEither(
+  asObject({
+    status: asValue('ok'),
+    msg: asString,
+    data: asObject({
+      amount: asNumber,
+      name: asString,
+      key: asString,
+      timeout: asNumber
+    })
+  }),
+  asObject({
+    status: asValue('failed'),
+    msg: asString
+  })
+)
+
+const asSmartPaySwapQuote = asObject({
+  status: asValue('ok'),
+  msg: asString,
+  data: asObject({
+    amount_usd: asString, // '0.000020',
+    price_brl: asString, // '0.1923',
+    total_brl: asString, // '0.00',
+    fee_brl: asString, // '0.00',
+    send_brl: asString, // '0.0001',
+    timeout: asNumber, // 491,
+    amount_txusdt: asString, // '0.000020',
+    price_txusdt: asNumber, // 1,
+    value_usd: asString, // '0.000020',
+    total_txusdt: asString // '0.000020'
+  })
+})
+
+export const parsePixKey = async (
+  io: EdgeIo,
+  currencyInfo: EdgeCurrencyInfo,
+  code: string,
+  smartPayPublicAddress: string,
+  smartPayUserId: string
+): Promise<EdgeParsedUri | undefined> => {
+  const now = new Date()
+  if (code.length > 36) {
+    const crc = computeCRC(code.slice(0, -4))
+    if (!code.endsWith(crc)) {
+      return
+    }
+    try {
+      const qrcode = encodeURIComponent(code)
+      const decode = await cleanMultiFetch(
+        asSmartPayQrDecode,
+        ['https://connect.smartpay.com.vc'],
+        `api/pix/qrdecode?qrcode=${qrcode}`,
+        undefined,
+        undefined,
+        io.fetch
+      )
+      if (decode.status !== 'ok') {
+        throw new Error(decode.msg)
+      }
+      const { data: decodeData } = decode
+      const { amount, key, name, timeout: decodeTimeout } = decodeData
+      if (decodeTimeout > 0 && decodeTimeout < 120) {
+        throw new Error('ErrorPixExpired')
+      }
+
+      let nativeAmount: string | undefined
+      let expireDate: Date | undefined
+      if (amount !== 0) {
+        const paramsObj: QueryParams = {
+          type: 'buy',
+          profile: 'transfer',
+          currency: 'brl',
+          conv: 'txusdt',
+          target: 'amount',
+          user: smartPayUserId,
+          amount: amount === 0 ? 100 : amount
+        }
+        const params = makeQueryParams(paramsObj)
+
+        // Get swap quote
+        const quote = await cleanMultiFetch(
+          asSmartPaySwapQuote,
+          ['https://connect.smartpay.com.vc'],
+          `api/swapix/swapquote?${params}`,
+          undefined,
+          undefined,
+          io.fetch
+        )
+
+        const { data: quoteData } = quote
+        const { amount_txusdt: amountTxusdt, timeout: quoteTimeout } = quoteData
+        if (quoteTimeout > 0 && quoteTimeout < 120) {
+          throw new Error('ErrorPixExpired')
+        }
+
+        // Get USDT info
+        const token = currencyInfo.metaTokens.find(
```

Comment:
We are going to delete `metaTokens` soon. Can you give this guy access to `builtinTokens` instead?

Reply:
Yeah, it's possible. You can grab the `builtinTokens` in the tools constructor, assign it to `this`, and then it will be available in parseUri.